### PR TITLE
Advanced quantization parameters

### PR DESCRIPTION
**Pull request by alexsu52**
### Changes

- Introduce [AdvancedQuantizationParameters](https://github.com/openvinotoolkit/nncf/blob/8f706f755f4af5130f2a1ff250e146cc7b476341/nncf/quantization/advanved_parameters.py#L99) for fine-tuning the quantization algorithm
- Introduce [AdvancedAccuracyRestorerParameters](https://github.com/openvinotoolkit/nncf/blob/8f706f755f4af5130f2a1ff250e146cc7b476341/nncf/quantization/advanved_parameters.py#L126) for fine-tuning the accuracy restorer algorithm.
- extend `quantize` method signature:
```
def quantize(
    ...
    advanced_parameters: Optional[AdvancedQuantizationParameters] = None
)  -> TModel:
```
- extend `quantize_with_accuracy_control` method signature:
```
def quantize_with_accuracy_control(
    ...
    advanced_quantization_parameters: Optional[AdvancedQuantizationParameters] = None,
    advanced_accuracy_restorer_parameters: Optional[AdvancedAccuracyRestorerParameters] = None
) -> TModel:
```
- Added backend parameters. Usage example:
``` 
# quantize OpenvVINO model using POT API
advanced_parameters = AdvancedQuantizationParameters(backend_params = {'use_pot': True})
quantize(ov_model, calibration_dataset, advanced_parameters)
```
- Updated `calibrate.py`
### Reason for changes

Extend quantization API

### Related tickets

ref: 96509

### Tests

TBD

### CODE REVIEW

File: nncf/quantization/quantize.py [hunk 1]
```diff
@@ -26,14 +26,16 @@
...
+def quantize(
+    model: TModel,
+    calibration_dataset: Dataset,
+    preset: QuantizationPreset = QuantizationPreset.PERFORMANCE,
+    target_device: TargetDevice = TargetDevice.ANY,
+    subset_size: int = 300,
+    fast_bias_correction: bool = True,
+    model_type: Optional[ModelType] = None,
+    ignored_scope: Optional[IgnoredScope] = None,
+    advanced_parameters: Optional[AdvancedQuantizationParameters()] = None) -> TModel:
```
Comment (andrey-churkin) [inline nncf/quantization/quantize.py:38]:
```suggestion
    advanced_parameters: Optional[AdvancedQuantizationParameters] = None) -> TModel:
```

File: nncf/quantization/parameters.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+
+class OverflowFix(Enum):
+    ENABLE = 'enable'
+    DISABLE = 'disable'
+    FIRST_LAYER_ONLY = 'first_layer_only'
+
+
+@dataclass
+class WeightsEstimatorParameters():
```
Comment (daniil-lyakhov) [inline nncf/quantization/parameters.py:54]: This would be super useful for inplace stats functionality!

File: nncf/openvino/pot/qunatization/quantize.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+    config = {}
+    if statistics_collector_params.statistics_type is not None:
+        config['type'] = statistics_collector_params.statistics_type.value
+    if statistics_collector_params.aggregator_type is not None:
+        config['aggregator'] = statistics_collector_params.aggregator_type.value
+    if statistics_collector_params.clipping_value is not None:
+        config['clipping_value'] = statistics_collector_params.clipping_value
+    if statistics_collector_params.statistics_type in [
+            StatisticsType.QUANTILE, StatisticsType.ABS_QUANTILE]:
+        config['outlier_prob'] = statistics_collector_params.quantile_outlier_prob
```
Comment (daniil-lyakhov) [inline nncf/openvino/pot/qunatization/quantize.py:125]: @alexsu52, we will use this `statistic_collecto_params` for `ov_native` backend, right?

Comment (alexsu52) [reply nncf/openvino/pot/qunatization/quantize.py:125]: Yes, I suggest using POT notation. Do you have any concern?

Comment (daniil-lyakhov) [reply nncf/openvino/pot/qunatization/quantize.py:125]: No concerns, looks doable 👍

File: nncf/quantization/advanved_parameters.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+        result['initializer']['range'] = [
+            activations_init_range_config,
+            weights_init_range_config
+        ]
+    
+    if params.bias_correction_params.apply_for_all_nodes:
+        raise RuntimeError(
+            'apply_for_all_nodes parameter of the BiasCorrection algorithm does not '
+            'support in the legacy format')
```
Comment (daniil-lyakhov) [inline nncf/quantization/advanved_parameters.py:229]:
```suggestion
            'apply_for_all_nodes parameter of the BiasCorrection algorithm is not '
            'supported in the legacy format')
```

File: nncf/quantization/advanved_parameters.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    
+    if params.bias_correction_params.threshold is not None:
+        raise RuntimeError(
+            'threshold parameter of the BiasCorrection algorithm does not support in '
+            'the legacy format')
```
Comment (daniil-lyakhov) [inline nncf/quantization/advanved_parameters.py:234]:
```suggestion
            'threshold parameter of the BiasCorrection algorithm is not supported  in '
            'the legacy format')
```

File: nncf/quantization/algorithms/bias_correction/algorithm.py [hunk 6]
```diff
@@ -92,24 +61,46 @@ class BiasCorrection(Algorithm):
...
+        self.backend_params = backend_params
         self.nncf_graph = None
         self._backend_entity = None
         self._collected_stat_inputs = set()
         self._fp_inputs = {}
+        
+        if self.apply_for_all_nodes:
+            raise RuntimeError(
+                'BiasCorrection algorithm from the OpenVINO backend does not support '
+                'apply_for_all_nodes=True yet')
```
Comment (daniil-lyakhov) [inline nncf/quantization/algorithms/bias_correction/algorithm.py:102]: Is this backend-specific runtime error in common code put in right position?

File: nncf/quantization/algorithms/fast_bias_correction/algorithm.py [hunk 7]
```diff
@@ -77,22 +51,43 @@ class FastBiasCorrection(Algorithm):
...
+        self.apply_for_all_nodes = apply_for_all_nodes
+        self.inplace_statistics = inplace_statistics
+        self.backend_params = backend_params
         self.nncf_graph = None
         self._backend_entity = None
+        
+        if self.apply_for_all_nodes:
+            raise RuntimeError(
+                'FastBiasCorrection algorithm from the OpenVINO backend does not '
+                'support apply_for_all_nodes=True yet')
```
Comment (daniil-lyakhov) [inline nncf/quantization/algorithms/fast_bias_correction/algorithm.py:90]: Is this backend-specific runtime error in common code put in right position?

File: nncf/quantization/algorithms/min_max/algorithm.py [hunk 8]
```diff
@@ -212,21 +214,51 @@ def _set_backend_entity(self, model: TModel) -> None:
...
+            from nncf.quantization.algorithms.min_max.torch_backend import PTMinMaxAlgoBackend
             self._backend_entity = PTMinMaxAlgoBackend()
         else:
             raise RuntimeError('Cannot return backend-specific entity '
                                'because {} is not supported!'.format(model_backend))
 
+    def _get_range_estimator_parameters(self,
+                                        target_point: TargetPoint,
+                                        quantizer_config: QuantizerConfig) -> RangeEstimatorParameters:
+        """_summary_
```
Comment (daniil-lyakhov) [inline nncf/quantization/algorithms/min_max/algorithm.py:233]: Is this ok?

File: nncf/openvino/pot/qunatization/__init__.py
Comment (l-bat) [inline nncf/openvino/pot/qunatization/__init__.py:1]: Wrong dir name: `qunatization` -> `quantization`.

File: nncf/quantization/algorithms/post_training/algorithm.py [hunk 9]
```diff
@@ -11,156 +11,125 @@
...
+            fine-tuning the quantization algorithm
         """
-        self.algorithms = {MinMaxQuantization: MinMaxQuantizationParameters(
+        super().__init__()
+        self.algorithms = []
+
+        if advanced_parameters is None:
+            advanced_parameters = AdvancedQuantizationParameters()
+
+        min_max_qunatization = MinMaxQuantization(
```
Comment (l-bat) [inline nncf/quantization/algorithms/post_training/algorithm.py:86]:
```suggestion
        min_max_quantization = MinMaxQuantization(
```

Comment (alexsu52) [reply nncf/quantization/algorithms/post_training/algorithm.py:86]: done

File: nncf/openvino/pot/qunatization/__init__.py
Comment (alexsu52) [reply nncf/openvino/pot/qunatization/__init__.py:1]: done

File: tests/onnx/quantization/test_qdq_params_calculation.py [hunk 10]
```diff
@@ -15,13 +15,14 @@
 import numpy as np
 import onnx
 import pytest
+
+from nncf.onnx.graph.onnx_graph import ONNXGraph
+from nncf.quantization.advanved_parameters import AdvancedQuantizationParameters, OverflowFix
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply tests/onnx/quantization/test_qdq_params_calculation.py:20]: done

File: nncf/torch/quantization/quantize.py [hunk 11]
```diff
@@ -95,31 +100,33 @@ def _get_transformer_quantization_config(subset_size: int) -> Dict:
             '{re}.*matmul_1',
             '{re}.*__truediv__*',
         ],
-        'overflow_fix': 'disable',
+        'overflow_fix': 'first_layer_only',
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/torch/quantization/quantize.py:103]: yes

File: nncf/tensorflow/quantization/quantize.py [hunk 12]
```diff
@@ -62,14 +64,67 @@ def transform_fn(data_item):
         return iter(DataProvider(self._dataset.get_inference_data(), transform_fn))
 
 
+def _get_default_quantization_config(preset: QuantizationPreset,
+                                     subset_size: int) -> Dict[str, Any]:
+    """
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/tensorflow/quantization/quantize.py:69]: done

File: nncf/tensorflow/quantization/quantize.py [hunk 13]
```diff
@@ -22,12 +22,14 @@
 from nncf.config.structures import QuantizationRangeInitArgs
 from nncf.data import Dataset
 from nncf.data.dataset import DataProvider
+from nncf.quantization.advanved_parameters import AdvancedQuantizationParameters, convert_advanced_parameters_to_dict
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/tensorflow/quantization/quantize.py:25]: done

File: nncf/quantization/advanved_parameters.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+    for field in fields(params):
+        value = getattr(params, field.name)
+        if value is not None:
+            changes[field.name] = value
+    return changes
+
+
+def convert_quantization_parameters_to_dict(params: QuantizationParameters) -> Dict[str, Any]:
+    """
+    Converts qunatization parameters to the dict in the legacy format
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/quantization/advanved_parameters.py]: done

File: nncf/quantization/advanved_parameters.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+
+    :param params: Qunatization parameters
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/quantization/advanved_parameters.py:127]: done

File: nncf/quantization/advanved_parameters.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+
+from dataclasses import dataclass
+from dataclasses import field
+from dataclasses import fields
+from enum import Enum
+import sys
+from typing import Any, ClassVar, Dict, Optional, Protocol
+
+from nncf.common.quantization.structs import QuantizationMode
+from nncf.quantization.range_estimator import AggregatorType, RangeEstimatorParameters, StatisticsType
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/quantization/advanved_parameters.py:22]: done

File: nncf/quantization/advanved_parameters.py [hunk 17]
```diff
@@ -0,0 +1,236 @@
+"""
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/quantization/advanved_parameters.py:1]: done

File: nncf/openvino/pot/qunatization/quantize.py [hunk 18]
```diff
@@ -0,0 +1,476 @@
...
+    backend_weights_parameters = advanced_parameters.backend_params.get(
+        BackendParameters.WEIGHTS, {})
+    weights_config = _create_quantization_group_config(
+        advanced_parameters.weights_quantization_params, advanced_parameters.
+        weights_range_estimator_params, backend_weights_parameters)
+    if weights_config:
+        config['weights'] = weights_config
+
+    if advanced_parameters.overflow_fix == OverflowFix.FIRST_LAYER:
+        config['saturation_fix'] = True
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/openvino/pot/qunatization/quantize.py]: done

File: nncf/experimental/openvino_native/quantization/quantize.py [hunk 19]
```diff
@@ -11,79 +11,106 @@
...
+                           'OpenVINO backend does not support the option to convert '
+                           'to the mixed preset yet')
+    
+    compress_weights = is_weight_compression_needed(advanced_quantization_parameters)
+    
+    if advanced_quantization_parameters is None:
+        copied_parameters = AdvancedQuantizationParameters()
+    else:
+        copied_parameters = deepcopy(advanced_quantization_parameters)
+    copied_parameters.backend_params[BackendParameters.COMPRESS_WEIGHTS] = False
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/experimental/openvino_native/quantization/quantize.py:109]: I set `False` because the default quantization should be called with `compress_weights = False`. https://github.com/openvinotoolkit/nncf/blob/7f09a5d9c5678f22cab35680133fcecf67296401/nncf/experimental/openvino_native/quantization/quantize.py#L111

File: nncf/quantization/algorithms/min_max/algorithm.py (hunk 8 above)
Comment (alexsu52) [reply nncf/quantization/algorithms/min_max/algorithm.py:233]: nope, fixed!

File: nncf/quantization/algorithms/fast_bias_correction/algorithm.py (hunk 7 above)
Comment (alexsu52) [reply nncf/quantization/algorithms/fast_bias_correction/algorithm.py:90]: This is not backend-specific runtime error. Fixed message.

File: nncf/quantization/algorithms/bias_correction/algorithm.py (hunk 6 above)
Comment (alexsu52) [reply nncf/quantization/algorithms/bias_correction/algorithm.py:102]: This is not backend-specific runtime error. Fixed message.

File: nncf/quantization/advanved_parameters.py (hunk 5 above)
Comment (alexsu52) [reply nncf/quantization/advanved_parameters.py:234]: done

File: nncf/quantization/advanved_parameters.py (hunk 4 above)
Comment (alexsu52) [reply nncf/quantization/advanved_parameters.py:229]: done

**Comment by alexsu52** [comment]
@daniil-lyakhov, @KodiaqQ, @l-bat thanks for your comments, please, review once more.

File: tests/openvino/pot/quantization/test_sanity.py [hunk 20]
```diff
@@ -25,7 +25,7 @@
 
 OMZ_MODELS = [
     ('resnet-18-pytorch', 'imagenette2-320', {'accuracy@top1': 0.777, 'accuracy@top5': 0.949}),
-    ('mobilenet-v3-small-1.0-224-tf', 'imagenette2-320', {'accuracy@top1': 0.737, 'accuracy@top5': 0.923}),
+    ('mobilenet-v3-small-1.0-224-tf', 'imagenette2-320', {'accuracy@top1': 0.744, 'accuracy@top5': 0.922}),
```
Comment (alexsu52) [inline tests/openvino/pot/quantization/test_sanity.py:28]: `inplace_statistics = True` is changed metrics of the mobilenet-v3-small-1.0-224-tf. As far I know `True` is default value for POT `inplace_statistics`, @KodiaqQ could you comment why feeding the default value is influence on result metrics of the quantized model?

**Review by l-bat — APPROVED** [review]

**Review by KodiaqQ — APPROVED** [review]

**Review by daniil-lyakhov — APPROVED** [review]
